File: frontend/src/domain/sourceEncoding.ts
export const ENCODING_CODECS = [
  "utf-8",
  "utf-16",
  "utf-16-le",
  "utf-16-be",
  "latin-1",
  "cp1251",
  "cp1252",
  "koi8-r",
  "cp866",
] as const;
export type EncodingCodec = (typeof ENCODING_CODECS)[number];
export type EncodingChoice = { readonly field_id: number } & (
  | {
      readonly mode: "keep" | "original";
      readonly encode_codec?: null;
      readonly decode_codec?: null;
    }
  | {
      readonly mode: "decode" | "codec";
      readonly encode_codec?: null;
      readonly decode_codec: EncodingCodec;
    }
  | {
      readonly mode: "unicode";
      readonly encode_codec: EncodingCodec;
      readonly decode_codec: EncodingCodec;
    }
);
export type EncodingField = {
  readonly field_id: number;
  readonly tag_name: string;
  readonly container: string;
  readonly physical_id: string | null;
  readonly extraction_version: string | null;
  readonly selected: boolean;
  readonly original_value: string;
  readonly current_value: string;
  readonly raw_evidence_available: boolean;
  readonly declared_codec: string | null;
  readonly applied_choice: EncodingChoice | null;
  readonly decision_origin?: "manual" | "auto" | null;
};
export type EncodingDetail = {
  readonly source_id: string;
  readonly source_revision: number;
  readonly fields: readonly EncodingField[];
  readonly suggestions?: readonly {
    readonly field_id: number;
    readonly state: "suggested" | "review";
    readonly value: string | null;
    readonly choice: EncodingChoice | null;
    readonly reason: string;
  }[];
};
export type EncodingRequest = {
  readonly expected_revision: number;
  readonly choices: readonly EncodingChoice[];
};
export type EncodingPreview = {
  readonly source_id: string;
  readonly source_revision: number;
  readonly valid: boolean;
  readonly fields: readonly {
    readonly field_id: number;
    readonly value: string | null;
    readonly status: "unchanged" | "changed" | "error";
    readonly error: string | null;
  }[];
};
export type EncodingApplied = EncodingPreview & { readonly queued: boolean };

export function isEncodingPreview(value: unknown): value is EncodingPreview {
  if (typeof value !== "object" || value === null) return false;
  return (
    "source_id" in value &&
    typeof value.source_id === "string" &&
    "source_revision" in value &&
    typeof value.source_revision === "number" &&
    "valid" in value &&
    typeof value.valid === "boolean" &&
    "fields" in value &&
    Array.isArray(value.fields) &&
    value.fields.every(
      (field: unknown) =>
        typeof field === "object" &&
        field !== null &&
        "field_id" in field &&
        typeof field.field_id === "number" &&
        "value" in field &&
        (field.value === null || typeof field.value === "string") &&
        "status" in field &&
        ["unchanged", "changed", "error"].includes(String(field.status)) &&
        "error" in field &&
        (field.error === null || typeof field.error === "string"),
    )
  );
}

File: src/music_ingest/contracts/source_encoding.py
from typing import Literal, Self

from pydantic import BaseModel, ConfigDict, Field, model_validator

Codec = Literal['utf-8', 'utf-16', 'utf-16-le', 'utf-16-be', 'latin-1', 'cp1251', 'cp1252', 'koi8-r', 'cp866']


class EncodingChoice(BaseModel):
    model_config = ConfigDict(extra='forbid')
    field_id: int = Field(gt=0)
    mode: Literal['keep', 'original', 'decode', 'unicode', 'codec'] = 'keep'
    encode_codec: Codec | None = None
    decode_codec: Codec | None = None

    @model_validator(mode='after')
    def valid_chain(self) -> Self:
        if self.mode in {'keep', 'original'} and (self.encode_codec or self.decode_codec):
            raise ValueError('keep/original do not accept codecs')
        if self.mode in {'decode', 'codec'} and (self.encode_codec is not None or self.decode_codec is None):
            raise ValueError('decode requires only decode_codec')
        if self.mode == 'unicode' and (self.encode_codec is None or self.decode_codec is None):
            raise ValueError('unicode requires encode_codec and decode_codec')
        return self


class EncodingRequest(BaseModel):
    model_config = ConfigDict(extra='forbid')
    expected_revision: int = Field(ge=1)
    choices: list[EncodingChoice] = Field(min_length=1, max_length=256)

    @model_validator(mode='after')
    def unique_fields(self) -> Self:
        if len({choice.field_id for choice in self.choices}) != len(self.choices):
            raise ValueError('duplicate field_id')
        return self


class EncodingPreviewField(BaseModel):
    field_id: int
    value: str | None
    status: Literal['unchanged', 'changed', 'error']
    error: str | None = None


class EncodingPreview(BaseModel):
    source_id: str
    source_revision: int
    valid: bool
    fields: list[EncodingPreviewField]


class EncodingField(BaseModel):
    field_id: int
    tag_name: str
    container: str
    physical_id: str | None
    extraction_version: str | None
    selected: bool
    original_value: str
    current_value: str
    raw_evidence_available: bool
    declared_codec: str | None
    applied_choice: EncodingChoice | None
    decision_origin: str | None = None


class EncodingSuggestion(BaseModel):
    field_id: int
    state: Literal['suggested', 'review']
    value: str | None
    choice: EncodingChoice | None
    reason: str


class EncodingDetail(BaseModel):
    source_id: str
    source_revision: int
    fields: list[EncodingField]
    suggestions: list[EncodingSuggestion] = Field(default_factory=list)


class EncodingApplied(EncodingPreview):
    queued: bool
